### PR TITLE
Evaluation metrics now stored in output folder

### PR DESCRIPTION
**Pull request by arangatang**
*Issue #, if available:*
#937
*Description of changes:*
aggregate metrics and item metrics are now stored in the output folder where the model is stored.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by arangatang** [comment]
When running the shell in sagemaker, this resolves too `/opt/ml/model` which according to the SageMaker [documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo-envvariables.html)

*`/opt/ml/model` —Your algorithm should write all final model artifacts to this directory. Amazon SageMaker copies this data as a single object in compressed tar format to the S3 location that you specified in the CreateTrainingJob request. If multiple containers in a single training job write to this directory they should ensure no file/directory names clash. Amazon SageMaker aggregates the result in a tar file and uploads to s3.*

**Comment by AaronSpieler** [comment]
> When running the shell in sagemaker, this resolves too `/opt/ml/model` which according to the SageMaker [documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo-envvariables.html)
> 
> _`/opt/ml/model` —Your algorithm should write all final model artifacts to this directory. Amazon SageMaker copies this data as a single object in compressed tar format to the S3 location that you specified in the CreateTrainingJob request. If multiple containers in a single training job write to this directory they should ensure no file/directory names clash. Amazon SageMaker aggregates the result in a tar file and uploads to s3._

Thats why I wouldn't put it there. I would suggest `/opt/ml/output/` or a subfolder like `/opt/ml/output/evaluation`.

**Comment by jaheba** [comment]
> > When running the shell in sagemaker, this resolves too `/opt/ml/model` which according to the SageMaker [documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo-envvariables.html)
> > _`/opt/ml/model` —Your algorithm should write all final model artifacts to this directory. Amazon SageMaker copies this data as a single object in compressed tar format to the S3 location that you specified in the CreateTrainingJob request. If multiple containers in a single training job write to this directory they should ensure no file/directory names clash. Amazon SageMaker aggregates the result in a tar file and uploads to s3._
> 
> Thats why I wouldn't put it there. I would suggest `/opt/ml/output/` or a subfolder like `/opt/ml/output/evaluation`.

I generally agree. But, I think the issue is that only the model folder is uploaded to S3, not the contents of `output`.

File: src/gluonts/shell/train.py [hunk 1]
```diff
@@ -133,3 +136,9 @@ def run_test(
     # we only log aggregate metrics for now as item metrics may be very large
     for name, score in agg_metrics.items():
         logger.info(f"#test_score ({env.current_host}, {name}): {score}")
+
+    # store metrics
+    with open(env.path.model / "agg_metrics.json", "w") as f:
+        json.dump(agg_metrics, f)
+    with open(env.path.model / "item_metrics.csv", "w") as f:
```
[the remark this answers is not in the document]
Comment (AaronSpieler) [reply src/gluonts/shell/train.py:143]: Saving them as a `.csv` makes it clear that it's a table. So for example `pd.read_csv('item_metrics.csv')` works great.

**Comment by AaronSpieler** [comment]
> > > When running the shell in sagemaker, this resolves too `/opt/ml/model` which according to the SageMaker [documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo-envvariables.html)
> > > _`/opt/ml/model` —Your algorithm should write all final model artifacts to this directory. Amazon SageMaker copies this data as a single object in compressed tar format to the S3 location that you specified in the CreateTrainingJob request. If multiple containers in a single training job write to this directory they should ensure no file/directory names clash. Amazon SageMaker aggregates the result in a tar file and uploads to s3._
> > 
> > 
> > Thats why I wouldn't put it there. I would suggest `/opt/ml/output/` or a subfolder like `/opt/ml/output/evaluation`.
> 
> I generally agree. But, I think the issue is that only the model folder is uploaded to S3, not the contents of `output`.

Interesting. So in the case of SageMaker Framework thats not the case. Anything written to the  `/opt/ml/output/` gets copied to S3 as well as, which is why for example in GluonTSFramework the training script looks like:

```
# input data, output dir and model directories
parser.add_argument(
    "--model-dir", type=str, default=os.environ["SM_MODEL_DIR"]
)
parser.add_argument(
    "--output-data_dir", type=str, default=os.environ["SM_OUTPUT_DATA_DIR"]
)
```

So unless the backend of what the shell uses works differently, I would expect `/opt/ml/output/` to be copied to S3 as well. Could we investigate this?

**Comment by jaheba** [comment]
> > > > When running the shell in sagemaker, this resolves too `/opt/ml/model` which according to the SageMaker [documentation](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo-envvariables.html)
> > > > _`/opt/ml/model` —Your algorithm should write all final model artifacts to this directory. Amazon SageMaker copies this data as a single object in compressed tar format to the S3 location that you specified in the CreateTrainingJob request. If multiple containers in a single training job write to this directory they should ensure no file/directory names clash. Amazon SageMaker aggregates the result in a tar file and uploads to s3._
> > > 
> > > 
> > > Thats why I wouldn't put it there. I would suggest `/opt/ml/output/` or a subfolder like `/opt/ml/output/evaluation`.
> > 
> > 
> > I generally agree. But, I think the issue is that only the model folder is uploaded to S3, not the contents of `output`.
> 
> Interesting. So in the case of SageMaker Framework thats not the case. Anything written to the `/opt/ml/output/` gets copied to S3 as well as, which is why for example in GluonTSFramework the training script looks like:
> 
> ```
> # input data, output dir and model directories
> parser.add_argument(
>     "--model-dir", type=str, default=os.environ["SM_MODEL_DIR"]
> )
> parser.add_argument(
>     "--output-data_dir", type=str, default=os.environ["SM_OUTPUT_DATA_DIR"]
> )
> ```
> 
> So unless the backend of what the shell uses works differently, I would expect `/opt/ml/output/` to be copied to S3 as well. Could we investigate this?

SageMaker Framework is not SageMaker. There is no option in SageMaker training jobs to specify an output-path, except the one which is used by models.

**Pull request merged**